drivers: serial: uart_sam0: Add RS485 transceiver GPIO support

rlevano77 opened the pull request:
Add RS485 half-duplex communication support to the SAM0 UART driver
through optional GPIO-controlled transceiver enable pins.

Changes:
- Add `de-gpios` and `re-gpios` properties to device tree binding for
  controlling RS485 transceiver Driver Enable and Receiver Enable pins
- Implement GPIO control in poll, interrupt-driven, and async modes
- Add RS485 flow control mode validation in runtime configuration
- Automatically manage transmitter/receiver enable state during TX/RX
  operations
- Initialize transceivers in receive mode by default

The implementation handles transceivers with separate or tied DE/RE pins
(specify only de-gpios when tied together). GPIO polarity is configurable
via device tree flags (GPIO_ACTIVE_HIGH/LOW).

Tested on: same54_xpro board

nandojve commented:
Hi @rlevano77 ,

This "Merge remote-tracking branch 'upstream/main" is not acceptable in Zephyr code base. You need to rebase on top of main only your commits.

This the merge commit, compliance checks and the tests so we can start review your PR.

rlevano77 commented:
Hi @nandojve I have rebased my commit on top of main and removed the merge commit. 
The PR now contains only RS485 changes with a clean linear history.
Please let me know if this is correct. Thank you!

nandojve commented on drivers/serial/uart_sam0.c:
```diff
@@ -54,6 +56,9 @@ struct uart_sam0_dev_cfg {
 	uint8_t rx_dma_channel;
 #endif
 	const struct pinctrl_dev_config *pcfg;
+	/* RS485 transceiver control (optional) */
```
Add 1 line space between lines 58 and 59

nandojve commented on dts/bindings/serial/atmel,sam0-uart.yaml:
```diff
@@ -1,4 +1,5 @@
 # Copyright (c) 2024 Gerson Fernando Budke <nandojve@gmail.com>
+# Copyright (c) 2025 Altronix (RS485 GPIO support)
```
```diff
- # Copyright (c) 2025 Altronix (RS485 GPIO support)
+ # Copyright (c) 2025 Altronix
```

nandojve reviewed (changes requested):
Hi @rlevano77 ,

Interesting work! Do you think it is possible to have a test ?
How do you suggest to validate code ?

rlevano77 committed:
dts: bindings: atmel,sam0: add RS485 GPIO properties

Add de-gpios and re-gpios properties to support RS485 transceiver
control. These optional GPIOs control Driver Enable (DE) and
Receiver Enable (RE) pins for half-duplex RS485 communication.

Signed-off-by: Ricardo Levano <rlevano77@gmail.com>

rlevano77 committed:
drivers: serial: sam0: add RS485 GPIO support

Add optional RS485 transceiver control using de-gpios and re-gpios
device tree properties. The implementation includes:

- Conditional compilation using DT_ANY_INST_HAS_PROP_STATUS_OKAY
- Helper functions for DE/RE GPIO control
- Integration with UART poll_out and interrupt handlers
- Runtime configuration validation for RS485 flow control
- Zero flash/RAM overhead when GPIOs not configured

Testing: Verified on SAME54 Xpro board with RS485 transceiver using
Zephyr's samples/drivers/uart/echo_bot. Setup included RS-485 breakout
board and USB-RS485 adapter. Tested with and without RS485 support.

Signed-off-by: Ricardo Levano <rlevano77@gmail.com>

rlevano77 committed:
samples: drivers: uart: add RS485 echo bot sample

Add a sample application demonstrating RS485 transceiver control
using the SAM0 UART driver. The sample shows how to:

- Configure UART in RS485 mode at runtime
- Use de-gpios and re-gpios for transceiver control
- Implement echo bot with interrupt-driven RX and polling TX

The sample includes complete documentation and overlay for
SAME54 Xplained Pro board using SERCOM0 with DE/RE control
on PA22.

Hardware tested:
- SAME54 Xplained Pro board
- SparkFun RS-485 Transceiver Breakout
- DTECH USB-to-RS485 converter

Signed-off-by: Ricardo Levano <rlevano77@gmail.com>

nandojve reviewed (dismissed):
Very well done!

nandojve commented:
Hi @rlevano77 , can you check CI issues ?

rlevano77 commented:
Hi @nandojve, I checked the CI issues. The failures on `twister-build (1)` and `twister-build (7)` are infrastructure errors — both show:

> "Executing the custom container implementation failed. Please contact your self hosted runner administrator."

Full logs here:
- [twister-build (1)](https://github.com/zephyrproject-rtos/zephyr/actions/runs/19959639677/job/57274597373)
- [twister-build (7)](https://github.com/zephyrproject-rtos/zephyr/actions/runs/19959639677/job/57274597379)

Builds 2–6 all passed successfully. The code itself is fine. Could you re-run the failed jobs, or should I rebase to trigger a fresh CI run?

pdgendt commented:
@rlevano77 there are build issues, try

```shell
west build -p -b arduino_zero/samd21g18a samples/synchronization -T sample.kernel.synchronization
```

rlevano77 commented:
@pdgendt Good catch — the `cfg` and `regs` variables in `uart_sam0_isr()` were moved to function scope but are only needed in the RS485 and async paths. This causes `-Werror=unused-variable` on boards without RS485 GPIOs configured (like `arduino_zero`). Fixed by scoping them to the blocks where they're used. Force-pushed with the fix.

pdgendt commented on dts/bindings/serial/atmel,sam0-uart.yaml:
Please add an example here (using the new `examples` property) like you did in the overlay file.

JarmouniA commented on dts/bindings/serial/atmel,sam0-uart.yaml:
```diff
@@ -129,3 +130,31 @@ properties:
+  de-gpios:
+    type: phandle-array
+    description: |
+      RS485 Driver Enable GPIO. Controls transmitter enable on RS485 transceiver.
```
```suggestion
      RS485  device Enable GPIO. Controls transmitter enable on RS485 transceiver.
```
#103087
